src/DIRAC/WorkloadManagementSystem/DB/VirtualMachineDB.py

```diff
@@ -0,0 +1,1435 @@
+
+  def getHistoryValues(self, averageBucket, selDict=None, fields2Get=False, timespan=0):
+    if not selDict:
+      selDict = {}
+    try:
+      timespan = max(0, int(timespan))
+    except ValueError:
+      return S_ERROR("Timespan has to be an integer")
+
+    cumulativeFields = ['Jobs', 'TransferredFiles', 'TransferredBytes']
+    validDataFields = ['Load', 'Jobs', 'TransferredFiles', 'TransferredBytes']
+    allValidFields = VirtualMachineDB.tablesDesc['vm_History']['Fields']
+
+    if not fields2Get:
+      fields2Get = list(validDataFields)
+    for field in fields2Get:
+      if field not in validDataFields:
+        return S_ERROR("%s is not a valid data field" % field)
+
+    # paramFields = fields2Get
+    try:
+      bucketSize = int(averageBucket)
+    except ValueError:
+      return S_ERROR("Average bucket has to be an integer")
+
+    sqlGroup = "FROM_UNIXTIME(UNIX_TIMESTAMP( `Update` ) - UNIX_TIMESTAMP( `Update` ) mod %d)" % bucketSize
+    sqlFields = ['`InstanceID`', sqlGroup]  # + [ "SUM(`%s`)/COUNT(`%s`)" % ( f, f ) for f in fields2Get ]
+    for field in fields2Get:
+      if field in cumulativeFields:
+        sqlFields.append("MAX(`%s`)" % field)
+      else:
+        sqlFields.append("SUM(`%s`)/COUNT(`%s`)" % (field, field))
+
+    sqlGroup = "%s, InstanceID" % sqlGroup
+    paramFields = ['Update'] + fields2Get
+    sqlCond = []
+
+    for field in selDict:
+      if field not in allValidFields:
+        return S_ERROR("%s is not a valid field" % field)
+      value = selDict[field]
+      if not isinstance(value, (list, tuple)):
+        value = (value, )
+      value = [self._escapeString(str(v))['Value'] for v in value]
+      sqlCond.append("`%s` in (%s)" % (field, ", ".join(value)))
+    if timespan > 0:
+      sqlCond.append("TIMESTAMPDIFF( SECOND, `Update`, UTC_TIMESTAMP() ) < %d" % timespan)
+    sqlQuery = "SELECT %s FROM `vm_History`" % ", ".join(sqlFields)
+    if sqlCond:
+      sqlQuery += " WHERE %s" % " AND ".join(sqlCond)
+    sqlQuery += " GROUP BY %s ORDER BY `Update` ASC" % sqlGroup
+    result = self._query(sqlQuery)
+    if not result['OK']:
+      return result
+    dbData = result['Value']
+    # Need ext?
+    requireExtension = set()
+    for i in range(len(fields2Get)):
+      f = fields2Get[i]
+      if f in cumulativeFields:
+        requireExtension.add(i)
+
+    if requireExtension:
+      rDates = []
+      for row in dbData:
+        if row[1] not in rDates:
+          rDates.append(row[1])
+      vmData = {}
+      for row in dbData:
+        vmID = row[0]
+        if vmID not in vmData:
+          vmData[vmID] = {}
+        vmData[vmID][row[1]] = row[2:]
+      rDates.sort()
+
+      dbData = []
+      for vmID in vmData:
+        prevValues = []
+        for rDate in rDates:
+          if rDate not in vmData[vmID]:
+            if prevValues:
+              instValues = [rDate]
+              for i in range(len(prevValues)):
+                instValues.append(prevValues[i])
```

Comment:
what's the need of this?

Reply:
Updated to use extend to generate the same list without the loop.